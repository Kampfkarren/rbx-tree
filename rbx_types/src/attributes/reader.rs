// Thanks to Anaminus! https://github.com/RobloxAPI/rbxattr/blob/master/spec.md

use super::*;
use crate::{basic_types::*, brick_color::BrickColor, variant::Variant};
use std::{
    collections::HashMap,
    io::{self, Read},
};

fn read_u8<R: Read>(mut reader: R) -> io::Result<u8> {
    let mut bytes = [0u8; 1];
    reader.read_exact(&mut bytes)?;
    Ok(bytes[0])
}

fn read_i32<R: Read>(mut reader: R) -> io::Result<i32> {
    let mut bytes = [0u8; 4];
    reader.read_exact(&mut bytes)?;
    Ok(i32::from_le_bytes(bytes))
}

fn read_u32<R: Read>(mut reader: R) -> io::Result<u32> {
    let mut bytes = [0u8; 4];
    reader.read_exact(&mut bytes)?;
    Ok(u32::from_le_bytes(bytes))
}

fn read_f32<R: Read>(mut reader: R) -> io::Result<f32> {
    let mut bytes = [0u8; 4];
    reader.read_exact(&mut bytes)?;
    Ok(f32::from_le_bytes(bytes))
}

fn read_string<R: Read>(mut reader: R) -> io::Result<Vec<u8>> {
    let size = read_u32(&mut reader)? as usize;
    let mut characters = vec![0u8; size];
    reader.read_exact(&mut characters)?;
    Ok(characters)
}

fn read_color3<R: Read>(mut reader: R) -> io::Result<Color3> {
    Ok(Color3::new(
        read_f32(&mut reader)?,
        read_f32(&mut reader)?,
        read_f32(&mut reader)?,
    ))
}

fn read_udim<R: Read>(mut reader: R) -> io::Result<UDim> {
    Ok(UDim::new(read_f32(&mut reader)?, read_i32(&mut reader)?))
}

fn read_vector2<R: Read>(mut reader: R) -> io::Result<Vector2> {
    Ok(Vector2::new(read_f32(&mut reader)?, read_f32(&mut reader)?))
}

/// Reads through an attribute property (AttributesSerialize) and returns a map of attribute names -> values.
pub fn get_attributes<R: Read>(mut value: R) -> Result<HashMap<Vec<u8>, Variant>, AttributeError> {
    let size = read_u32(&mut value).map_err(|_| AttributeError::InvalidSize)?;
    let mut attributes = HashMap::with_capacity(size as usize);

    for _ in 0..size {
        let name = read_string(&mut value).map_err(|_| AttributeError::InvalidName)?;

        let attribute_key =
            AttributeType::try_from(read_u8(&mut value).map_err(|_| AttributeError::NoValueType)?)?;

        let value = match attribute_key {
            AttributeType::BrickColor => {
                let color =
                    read_u32(&mut value).map_err(|_| AttributeError::Other("BrickColor"))?;

                Variant::BrickColor(
                    BrickColor::from_number(color as u16)
                        .ok_or(AttributeError::InvalidBrickColor(color))?,
                )
            }

            AttributeType::Bool => {
                Variant::Bool(read_u8(&mut value).map_err(|_| AttributeError::Other("bool"))? != 0)
            }

            AttributeType::Color3 => Variant::Color3(
                read_color3(&mut value).map_err(|_| AttributeError::Other("Color3"))?,
            ),

            AttributeType::ColorSequence => {
                let size = read_u32(&mut value).map_err(|_| AttributeError::InvalidSize)?;
                let mut keypoints = Vec::with_capacity(size as usize);

                for _ in 0..size {
                    // `envelope` is not represented in rbx_types, apparently it's always zero.
                    let _envelope = read_f32(&mut value)
                        .map_err(|_| AttributeError::Other("ColorSequenceKeypoint envelope"))?;

                    let time = read_f32(&mut value)
                        .map_err(|_| AttributeError::Other("ColorSequenceKeypoint time"))?;

                    let color = read_color3(&mut value)
                        .map_err(|_| AttributeError::Other("ColorSequenceKeypoint color"))?;

                    keypoints.push(ColorSequenceKeypoint::new(time, color));
                }

                Variant::ColorSequence(ColorSequence { keypoints })
            }

            AttributeType::Float32 => {
                Variant::Float32(read_f32(&mut value).map_err(|_| AttributeError::Other("float"))?)
            }

            AttributeType::Float64 => {
                let mut bytes = [0u8; 8];
                value
                    .read_exact(&mut bytes)
                    .map_err(|_| AttributeError::Other("double"))?;
                Variant::Float64(f64::from_le_bytes(bytes))
            }

            AttributeType::NumberRange => Variant::NumberRange(NumberRange::new(
                read_f32(&mut value).map_err(|_| AttributeError::Other("NumberRange min"))?,
                read_f32(&mut value).map_err(|_| AttributeError::Other("NumberRange max"))?,
            )),

            AttributeType::NumberSequence => {
                let size = read_u32(&mut value).map_err(|_| AttributeError::InvalidSize)?;

                let mut keypoints = Vec::with_capacity(size as usize);

                for _ in 0..size {
                    let envelope = read_f32(&mut value)
                        .map_err(|_| AttributeError::Other("NumberSequence envelope"))?;

                    let time = read_f32(&mut value)
                        .map_err(|_| AttributeError::Other("NumberSequence time"))?;

                    let value = read_f32(&mut value)
                        .map_err(|_| AttributeError::Other("NumberSequence value"))?;

                    keypoints.push(NumberSequenceKeypoint::new(time, value, envelope));
                }

                Variant::NumberSequence(NumberSequence { keypoints })
            }

            AttributeType::Rect => Variant::Rect(Rect::new(
                read_vector2(&mut value).map_err(|_| AttributeError::Other("Rect min"))?,
                read_vector2(&mut value).map_err(|_| AttributeError::Other("Rect max"))?,
            )),

            AttributeType::BinaryString => Variant::BinaryString(
                read_string(&mut value)
                    .map_err(|_| AttributeError::Other("string"))?
                    .into(),
            ),

            AttributeType::UDim => {
                Variant::UDim(read_udim(&mut value).map_err(|_| AttributeError::Other("UDim"))?)
            }

            AttributeType::UDim2 => Variant::UDim2(UDim2::new(
                read_udim(&mut value).map_err(|_| AttributeError::Other("UDim2 X"))?,
                read_udim(&mut value).map_err(|_| AttributeError::Other("UDim2 Y"))?,
            )),

            AttributeType::Vector2 => Variant::Vector2(Vector2::new(
                read_f32(&mut value).map_err(|_| AttributeError::Other("Vector2 X"))?,
                read_f32(&mut value).map_err(|_| AttributeError::Other("Vector2 Y"))?,
            )),

            AttributeType::Vector3 => Variant::Vector3(Vector3::new(
                read_f32(&mut value).map_err(|_| AttributeError::Other("Vector3 X"))?,
                read_f32(&mut value).map_err(|_| AttributeError::Other("Vector3 Y"))?,
                read_f32(&mut value).map_err(|_| AttributeError::Other("Vector3 Z"))?,
            )),
        };

        attributes.insert(name, value);
    }

    Ok(attributes)
}
