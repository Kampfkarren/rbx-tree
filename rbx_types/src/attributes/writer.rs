use super::*;
use crate::{basic_types::*, variant::Variant};
use std::io::Write;

fn write_f32<W: Write>(mut writer: W, n: f32) {
    writer
        .write_all(&mut n.to_le_bytes().to_vec())
        .expect("couldn't write f32 to buffer");
}

fn write_u32<W: Write>(mut writer: W, n: u32) {
    writer
        .write_all(&mut n.to_le_bytes().to_vec())
        .expect("couldn't write u32 to buffer");
}

fn write_color3<W: Write>(mut writer: W, color: Color3) {
    write_f32(&mut writer, color.r);
    write_f32(&mut writer, color.g);
    write_f32(&mut writer, color.b);
}

fn write_string<W: Write>(mut writer: W, string: &str) {
    write_u32(&mut writer, string.len() as u32);
    writer
        .write_all(&string.bytes().collect::<Vec<_>>())
        .expect("couldn't write string to buffer");
}

fn write_udim<W: Write>(mut writer: W, udim: UDim) {
    write_f32(&mut writer, udim.scale);
    writer
        .write_all(&mut udim.offset.to_le_bytes().to_vec())
        .expect("couldn't write UDim to buffer");
}

fn write_vector2<W: Write>(mut writer: W, vector2: Vector2) {
    write_f32(&mut writer, vector2.x);
    write_f32(&mut writer, vector2.y);
}

/// Writes the attribute property (AttributesSerialize) from a map of attribute names -> values.
pub fn attributes_from_map<
    K: Into<String>,
    V: Into<Variant>,
    I: Iterator<Item = (K, V)> + ExactSizeIterator,
    M: IntoIterator<IntoIter = I, Item = (K, V)>,
>(
    map: M,
) -> Result<Vec<u8>, AttributeError> {
    let map = map.into_iter();
    let mut bytes = Vec::new();

    bytes
        .write_all(&(map.len() as u32).to_le_bytes())
        .expect("couldn't write map length to buffer");

    for (name, variant) in map {
        let variant = variant.into();

        write_string(&mut bytes, &name.into());

        let attribute_type = AttributeType::try_from(variant.ty())?;
        bytes.push(attribute_type as u8);

        match (attribute_type, variant) {
            (AttributeType::Bool, Variant::Bool(bool)) => bytes.push(bool as u8),
            (AttributeType::BrickColor, Variant::BrickColor(color)) => {
                write_u32(&mut bytes, color as u32)
            }
            (AttributeType::Color3, Variant::Color3(color)) => write_color3(&mut bytes, color),
            (AttributeType::ColorSequence, Variant::ColorSequence(sequence)) => {
                write_u32(&mut bytes, sequence.keypoints.len() as u32);

                for keypoint in &sequence.keypoints {
                    write_f32(&mut bytes, 0.0); // Envelope
                    write_f32(&mut bytes, keypoint.time);
                    write_color3(&mut bytes, keypoint.color);
                }
            }
            (AttributeType::Float32, Variant::Float32(float)) => {
                write_f32(&mut bytes, float);
            }
            (AttributeType::Float64, Variant::Float64(float)) => {
                bytes.append(&mut float.to_le_bytes().to_vec());
            }
            (AttributeType::NumberRange, Variant::NumberRange(range)) => {
                write_f32(&mut bytes, range.min);
                write_f32(&mut bytes, range.max);
            }
            (AttributeType::NumberSequence, Variant::NumberSequence(sequence)) => {
                write_u32(&mut bytes, sequence.keypoints.len() as u32);

                for keypoint in &sequence.keypoints {
                    write_f32(&mut bytes, keypoint.envelope);
                    write_f32(&mut bytes, keypoint.time);
                    write_f32(&mut bytes, keypoint.value);
                }
            }
            (AttributeType::Rect, Variant::Rect(rect)) => {
                write_vector2(&mut bytes, rect.min);
                write_vector2(&mut bytes, rect.max);
            }
            (AttributeType::String, Variant::String(string)) => {
                write_string(&mut bytes, &string);
            }
            (AttributeType::UDim, Variant::UDim(udim)) => {
                write_udim(&mut bytes, udim);
            }
            (AttributeType::UDim2, Variant::UDim2(udim2)) => {
                write_udim(&mut bytes, udim2.x);
                write_udim(&mut bytes, udim2.y);
            }
            (AttributeType::Vector2, Variant::Vector2(vector2)) => {
                write_vector2(&mut bytes, vector2);
            }
            (AttributeType::Vector3, Variant::Vector3(vector3)) => {
                write_f32(&mut bytes, vector3.x);
                write_f32(&mut bytes, vector3.y);
                write_f32(&mut bytes, vector3.z);
            }

            (other_type, other_variant) => unreachable!(
                "variant {:?} didn't match attribute type {:?}",
                other_variant, other_type
            ),
        }
    }

    Ok(bytes)
}
