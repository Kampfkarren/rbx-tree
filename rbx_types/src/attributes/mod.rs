use crate::variant::{self, VariantType};

use std::{
    borrow::Borrow,
    collections::HashMap,
    convert::TryFrom,
    io::{self, Read, Write},
    iter::FromIterator,
    string::FromUtf8Error,
};

use thiserror::Error;

use variant::Variant;

mod reader;
mod writer;

use reader::get_attributes;

use writer::attributes_from_map;

macro_rules! create_attribute_type {
    ({
        $(
            $key:ident = $number:tt,
        )+
    }) => {
        #[derive(Clone, Copy, Debug, Eq, PartialEq)]
        pub(crate) enum AttributeType {
            $(
                $key = $number,
            )+
        }

        impl TryFrom<VariantType> for AttributeType {
            type Error = AttributeError;

            fn try_from(variant_type: VariantType) -> Result<Self, Self::Error> {
                match variant_type {
                    $(
                        VariantType::$key => Ok(AttributeType::$key),
                    )+

                    _ => Err(AttributeError::InvalidVariantType),
                }
            }
        }

        impl TryFrom<u8> for AttributeType {
            type Error = AttributeError;

            fn try_from(byte: u8) -> Result<Self, Self::Error> {
                match byte {
                    $(
                        $number => Ok(Self::$key),
                    )+

                    other => Err(AttributeError::InvalidValueType(other))
                }
            }
        }
    };
}

create_attribute_type!({
    // ??? = 0x01,
    BinaryString = 0x02,
    Bool = 0x03,
    // ??? = 0x04,
    Float32 = 0x05,
    Float64 = 0x06,
    // ??? = 0x07,
    // ??? = 0x08,
    UDim = 0x09,
    UDim2 = 0x0A,
    // ??? = 0x0B,
    // ??? = 0x0C,
    // ??? = 0x0D,
    BrickColor = 0x0E,
    Color3 = 0x0F,
    Vector2 = 0x10,
    Vector3 = 0x11,
    // ??? = 0x12,
    // ??? = 0x13,
    // ??? = 0x14,
    // ??? = 0x15,
    // ??? = 0x16,
    NumberSequence = 0x17,
    // ??? = 0x18,
    ColorSequence = 0x19,
    // ??? = 0x1A,
    NumberRange = 0x1B,
    Rect = 0x1C,
});

#[derive(Debug, Default, Clone, PartialEq)]
pub struct Attributes {
    data: HashMap<String, Variant>,
}

impl Attributes {
    /// Creates an empty `Attributes` struct
    pub fn new() -> Self {
        Self::default()
    }

    /// Reads from a serialized attributes string, and produces a new `Attributes` from it.
    pub fn from_reader<R: Read>(reader: R) -> Result<Self, AttributeError> {
        Ok(Attributes {
            data: get_attributes(reader)?,
        })
    }

    /// Writes the attributes as a serialized string to the writer.
    pub fn to_writer<W: Write>(&self, mut writer: W) -> Result<(), AttributeError> {
        let bytes = attributes_from_map(self.data.iter())?;
        writer
            .write_all(&bytes)
            .map_err(AttributeError::ToWriterFail)
    }

    /// Get the attribute with the following key.
    pub fn get<K: Borrow<str>>(&self, key: K) -> Option<&Variant> {
        self.data.get(key.borrow())
    }

    /// Inserts an attribute with the given key and value.
    /// Will return the attribute that used to be there if one existed.
    pub fn insert(&mut self, key: String, value: Variant) -> Option<Variant> {
        self.data.insert(key, value)
    }

    /// Returns an iterator of borrowed attributes.
    pub fn iter(&self) -> impl Iterator<Item = (&String, &Variant)> {
        self.data.iter()
    }
}

impl IntoIterator for Attributes {
    type IntoIter = AttributesIntoIter;
    type Item = (String, Variant);

    fn into_iter(self) -> Self::IntoIter {
        AttributesIntoIter {
            iter: self.data.into_iter(),
        }
    }
}

impl FromIterator<(String, Variant)> for Attributes {
    fn from_iter<T: IntoIterator<Item = (String, Variant)>>(iter: T) -> Self {
        Self {
            data: iter.into_iter().collect(),
        }
    }
}

/// An owning iterator over the entries of an `Attributes`.
/// This is created by [`Attributes::into_iter`].
pub struct AttributesIntoIter {
    iter: std::collections::hash_map::IntoIter<String, Variant>,
}

impl Iterator for AttributesIntoIter {
    type Item = (String, Variant);

    fn next(&mut self) -> Option<Self::Item> {
        self.iter.next()
    }
}

#[derive(Debug, Error)]
pub enum AttributeError {
    #[error("invalid value type: {0}")]
    InvalidValueType(u8),

    #[error("invalid brick color: {0}")]
    InvalidBrickColor(u32),

    #[error("invalid entry key")]
    InvalidEntryKey,

    #[error("invalid name")]
    InvalidName,

    #[error("invalid size")]
    InvalidSize,

    #[error("invalid variant type passed")]
    InvalidVariantType,

    #[error("no value type was found")]
    NoValueType,

    #[error("malformed attribute key")]
    MalformedEntryKey(FromUtf8Error),

    #[error("couldn't write to writer: {0}")]
    ToWriterFail(io::Error),

    #[error("couldn't read bytes to deserialize {0}")]
    Other(&'static str),
}

#[cfg(all(feature = "serde", test))]
mod tests {
    use std::collections::BTreeMap;

    use super::*;

    // This is taken from rbx-test-files/models/attributes/xml.rbxmx.
    // This is pasted raw as to not create a circular dependency in test (rbx_types -> rbx_xml/rbx_binary -> rbx_types)
    const ATTRIBUTES_BASE64: &str = "\
    DwAAAAMAAABOYU4GAAAAAAAA+P8IAAAASW5maW5pdHkGAAAAAAAA8H8NAAAAQ29sb3JTZXF1\
    ZW5jZRkDAAAAAAAAAAAAAAAAAIA/AAAAAAAAAAAAAAAAAAAAPwAAAAAAAIA/AAAAAAAAAAAA\
    AIA/AAAAAAAAAAAAAIA/BwAAAFZlY3RvcjMRAACAPwAAAEAAAEBABwAAAFZlY3RvcjIQAAAg\
    QQAASEIOAAAATnVtYmVyU2VxdWVuY2UXAwAAAAAAAAAAAAAAAACAPwAAAAAAAAA/AAAAAAAA\
    AAAAAIA/AACAPwYAAABDb2xvcjMPo6IiPwAAAAAAAIA/CgAAAEJyaWNrQ29sb3IO7AMAAAQA\
    AABSZWN0HAAAgD8AAABAAABAQAAAgEAFAAAAVURpbTIKAAAAPwoAAAAzMzM/HgAAAAQAAABV\
    RGltCQAAAD9kAAAACwAAAE51bWJlclJhbmdlGwAAoEAAACBBBgAAAE51bWJlcgYAAAAAgBzI\
    QAcAAABCb29sZWFuAwEGAAAAU3RyaW5nAg0AAABIZWxsbywgd29ybGQh";

    #[test]
    fn test_round_trip_attributes() {
        let attributes_value =
            base64::decode(ATTRIBUTES_BASE64).expect("bad base64 for attributes");

        let attributes = Attributes::from_reader(&attributes_value[..])
            .expect("couldn't deserialize attributes");

        let attributes_stable_order = BTreeMap::from_iter(attributes.clone().into_iter());
        insta::assert_yaml_snapshot!(attributes_stable_order);

        let mut new_attribute_bytes = Vec::<u8>::new();
        attributes
            .to_writer(&mut new_attribute_bytes)
            .expect("couldn't write attributes to buffer");

        let byte_string = new_attribute_bytes
            .iter()
            .map(|byte| format!("{:x?}", byte))
            .collect::<String>();
        insta::assert_snapshot!(byte_string);

        let new_attributes = Attributes::from_reader(new_attribute_bytes.as_slice())
            .expect("couldn't deserialize crate produced binary");

        let new_attributes_stable_order = BTreeMap::from_iter(new_attributes.into_iter());

        // They are not checked directly against each other because the data contains NaN.
        assert_eq!(
            format!("{:#?}", attributes_stable_order),
            format!("{:#?}", new_attributes_stable_order)
        );
    }
}
