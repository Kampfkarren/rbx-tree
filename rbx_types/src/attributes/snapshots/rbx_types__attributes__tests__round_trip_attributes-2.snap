---
source: rbx_types/src/attributes/mod.rs
expression: byte_string
---
f00060004e756d62657260000801cc8404000526563741c00803f000400040400080406000537472696e672d00048656c6c6f2c20776f726c64218000496e66696e6974796000000f07fd000436f6c6f7253657175656e63651930000000000000803f0000000000000003f000000803f0000000000803f0000000000803f30004e614e6000000f8ff7000566563746f723210002041004842e0004e756d62657253657175656e63651730000000000000803f00000003f0000000000803f00803f6000436f6c6f7233fa3a2223f000000803fa000427269636b436f6c6f72eec30050005544696d32a0003fa0003333333f1e00040005544696d90003f640007000426f6f6c65616e317000566563746f72331100803f00040004040b0004e756d62657252616e67651b00a040002041
